File: src/main/java/com/example/tutorial/dsa/basics/arrays/DuplicateDataDetector.java
package com.example.tutorial.dsa.basics.arrays;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * DuplicateDataDetector
 * ----------------------------------
 * <p>This program detects duplicate elements in an array using HashSet.
 * The core problem solved here is Contains Duplicate (LeetCode #217).
 *
 * <p><b>Problem Statement:</b>
 * Given an integer array nums, return true if any value appears at least twice
 * in the array, and return false if every element is distinct.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Detect duplicate offer IDs in a batch upload</li>
 *   <li>Validate unique campaign assignments</li>
 *   <li>Check for duplicate merchant entries</li>
 * </ul>
 *
 * <p><b>Examples:</b>
 * <ul>
 *   <li>Input: [1,2,3,1] → Output: true (1 appears twice)</li>
 *   <li>Input: [1,2,3,4] → Output: false (all distinct)</li>
 *   <li>Input: [1,1,1,3,3,4,3,2,4,2] → Output: true</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon, Apple, Google
 *
 * @see <a href="https://leetcode.com/problems/contains-duplicate/">LeetCode 217 - Contains Duplicate</a>
 */
@Component
public class DuplicateDataDetector implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(DuplicateDataDetector.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== DuplicateDataDetector: Duplicate Detection Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate duplicate detection with offer IDs
    System.out.println("--- Checking for Duplicate Offer IDs ---\n");
    String[] offerIds = offers.stream()
        .map(Offer::getOfferId)
        .toArray(String[]::new);

    boolean hasDuplicateOffers = containsDuplicateStrings(offerIds);
    System.out.println("Offer IDs: " + java.util.Arrays.toString(offerIds));
    System.out.println("Contains duplicates: " + hasDuplicateOffers + "\n");

    // Demonstrate with numeric arrays
    System.out.println("--- Numeric Array Examples ---\n");
    int[][] testCases = {
        {1, 2, 3, 1},
        {1, 2, 3, 4},
        {1, 1, 1, 3, 3, 4, 3, 2, 4, 2}
    };

    for (int[] nums : testCases) {
      boolean hasDuplicate = containsDuplicate(nums);
      System.out.println("Array: " + java.util.Arrays.toString(nums));
      System.out.println("Contains duplicate: " + hasDuplicate + "\n");
    }
  }

  /**
   * Detects if array contains any duplicate elements using HashSet.
   *
   * <p><b>LOGIC:</b>
   * <ol>
   *   <li>Create a HashSet to store seen elements</li>
   *   <li>Iterate through each element in the array</li>
   *   <li>For each element, check if it's already in the set</li>
   *   <li>If yes, we found a duplicate - return true</li>
   *   <li>If no, add element to set and continue</li>
   *   <li>If we finish without finding duplicate, return false</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * nums = [1, 2, 3, 1]
   *
   * Step 1: num=1, set={}, 1 not in set → add 1, set={1}
   * Step 2: num=2, set={1}, 2 not in set → add 2, set={1,2}
   * Step 3: num=3, set={1,2}, 3 not in set → add 3, set={1,2,3}
   * Step 4: num=1, set={1,2,3}, 1 IS in set → return true!
   *
   * Result: true (duplicate found)
   * </pre>
   *
   * <p><b>Time Complexity: O(n)</b>
   * <br>We visit each element once, and HashSet operations (add/contains) are O(1).
   * <br><i>Like checking names off a guest list - each lookup is instant because
   * the list is organized alphabetically. With n guests, we do n quick lookups.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>In worst case (no duplicates), we store all n elements in the HashSet.
   * <br><i>Like writing down each guest's name - if all 100 guests are unique,
   * you need paper for 100 names. The space grows with input size.</i>
   *
   * @param nums the array of integers to check
   * @return true if any value appears at least twice, false otherwise
   */
  public static boolean containsDuplicate(int[] nums) {
    // Edge case: empty or single element array can't have duplicates.
    if (nums == null || nums.length <= 1) {
      return false;
    }

    // HashSet for O(1) lookup - like an indexed phonebook.
    Set<Integer> seen = new HashSet<>();

    // Check each element.
    for (int num : nums) {
      // If already seen, it's a duplicate!
      if (seen.contains(num)) {
        return true;
      }
      // Otherwise, remember we've seen this number.
      seen.add(num);
    }

    // No duplicates found.
    return false;
  }

  /**
   * String version for detecting duplicate strings (like offer IDs).
   *
   * @param strings the array of strings to check
   * @return true if any string appears at least twice
   */
  public static boolean containsDuplicateStrings(String[] strings) {
    if (strings == null || strings.length <= 1) {
      return false;
    }

    Set<String> seen = new HashSet<>();

    for (String str : strings) {
      if (seen.contains(str)) {
        return true;
      }
      seen.add(str);
    }

    return false;
  }
}

File: src/main/python/com/example/tutorial/dsa/basics/arrays/duplicate_data_detector.py
"""
DuplicateDataDetector
----------------------------------
This program detects duplicate elements in an array using HashSet.
The core problem solved here is Contains Duplicate (LeetCode #217).

Problem Statement:
    Given an integer array nums, return true if any value appears at least twice
    in the array, and return false if every element is distinct.

Real UseCase:
    In a credit card offers system:
    - Detect duplicate offer IDs in a batch upload
    - Validate unique campaign assignments
    - Check for duplicate merchant entries

Examples:
    - Input: [1,2,3,1] -> Output: True (1 appears twice)
    - Input: [1,2,3,4] -> Output: False (all distinct)
    - Input: [1,1,1,3,3,4,3,2,4,2] -> Output: True

Company Tags: Amazon, Apple, Google

See: https://leetcode.com/problems/contains-duplicate/
"""

import sys
import os
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


def contains_duplicate(nums: List[int]) -> bool:
    """
    Detects if array contains any duplicate elements using HashSet.

    LOGIC:
        1. Create a set to store seen elements
        2. Iterate through each element in the array
        3. For each element, check if it's already in the set
        4. If yes, we found a duplicate - return True
        5. If no, add element to set and continue
        6. If we finish without finding duplicate, return False

    Example Walkthrough:
        nums = [1, 2, 3, 1]

        Step 1: num=1, seen={}, 1 not in seen -> add 1, seen={1}
        Step 2: num=2, seen={1}, 2 not in seen -> add 2, seen={1,2}
        Step 3: num=3, seen={1,2}, 3 not in seen -> add 3, seen={1,2,3}
        Step 4: num=1, seen={1,2,3}, 1 IS in seen -> return True!

        Result: True (duplicate found)

    Time Complexity: O(n)
        We visit each element once, and set operations (add/in) are O(1).
        Like checking names off a guest list - each lookup is instant because
        the list is organized alphabetically. With n guests, we do n quick lookups.

    Space Complexity: O(n)
        In worst case (no duplicates), we store all n elements in the set.
        Like writing down each guest's name - if all 100 guests are unique,
        you need paper for 100 names. The space grows with input size.

    Args:
        nums: The array of integers to check.

    Returns:
        True if any value appears at least twice, False otherwise.
    """
    # Edge case: empty or single element array can't have duplicates.
    if not nums or len(nums) <= 1:
        return False

    # Set for O(1) lookup - like an indexed phonebook.
    seen = set()

    # Check each element.
    for num in nums:
        # If already seen, it's a duplicate!
        if num in seen:
            return True
        # Otherwise, remember we've seen this number.
        seen.add(num)

    # No duplicates found.
    return False


def contains_duplicate_strings(strings: List[str]) -> bool:
    """
    String version for detecting duplicate strings (like offer IDs).

    Args:
        strings: The array of strings to check.

    Returns:
        True if any string appears at least twice.
    """
    if not strings or len(strings) <= 1:
        return False

    seen = set()

    for s in strings:
        if s in seen:
            return True
        seen.add(s)

    return False


def main():
    """Main function to demonstrate the DuplicateDataDetector."""
    print("=== DuplicateDataDetector: Duplicate Detection Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate duplicate detection with offer IDs
    print("--- Checking for Duplicate Offer IDs ---\n")
    offer_ids = [offer.offer_id for offer in offers]

    has_duplicate_offers = contains_duplicate_strings(offer_ids)
    print(f"Offer IDs: {offer_ids}")
    print(f"Contains duplicates: {has_duplicate_offers}\n")

    # Demonstrate with numeric arrays
    print("--- Numeric Array Examples ---\n")
    test_cases = [
        [1, 2, 3, 1],
        [1, 2, 3, 4],
        [1, 1, 1, 3, 3, 4, 3, 2, 4, 2]
    ]

    for nums in test_cases:
        has_duplicate = contains_duplicate(nums)
        print(f"Array: {nums}")
        print(f"Contains duplicate: {has_duplicate}\n")


if __name__ == "__main__":
    main()
